File: frontend/src/features/dynamic-simulation/algorithms/round_robin.ts
/**
 * Round Robin — TypeScript port
 * Deterministic baseline cycling through all gateways.
 */

import { BaseAlgorithm } from './base';
import { TransactionContext, ArmState, AlgorithmMetadata, HyperparameterSchema } from '../types';

export class RoundRobin implements BaseAlgorithm {
    private arms: string[] = [];
    private stepCount = 0;
    private successes: Record<string, number> = {};
    private counts: Record<string, number> = {};

    initialize(arms: string[], _config: Record<string, number>): void {
        this.arms = arms;
        this.stepCount = 0;
        this.successes = {};
        this.counts = {};
        for (const arm of arms) {
            this.successes[arm] = 0;
            this.counts[arm] = 0;
        }
    }

    select(_context: TransactionContext): string {
        const chosen = this.arms[this.stepCount % this.arms.length];
        this.stepCount++;
        return chosen;
    }

    update(arm: string, reward: number, _context: TransactionContext): void {
        this.counts[arm] += 1;
        this.successes[arm] += reward;
    }

    getState(): Record<string, ArmState> {
        const state: Record<string, ArmState> = {};
        for (const arm of this.arms) {
            const n = this.counts[arm];
            state[arm] = {
                estimatedSR: n > 0 ? this.successes[arm] / n : null,
                selectionScore: null,
                totalSelections: n,
            };
        }
        return state;
    }

    explainLastDecision(): string {
        const idx = (this.stepCount - 1) % this.arms.length;
        return `Round Robin: selected '${this.arms[idx]}' at step ${this.stepCount}.`;
    }

    getHyperparameterSchema(): HyperparameterSchema {
        return {};
    }

    metadata(): AlgorithmMetadata {
        return {
            name: 'Round Robin',
            shortName: 'RR',
            description: 'Deterministic baseline cycling through all gateways. No outcome feedback.',
            paper: 'Deterministic baseline',
            paperUrl: '',
            category: 'rule_based',
        };
    }
}

File: algorithms/round_robin.py
"""
Round Robin baseline algorithm.
Deterministic cycling through all gateways regardless of outcome.
"""

from typing import List, Dict, Any
from algorithms.base import BaseAlgorithm, TransactionContext


class RoundRobin(BaseAlgorithm):
    """
    Deterministic baseline: cycles through gateways in fixed order.
    Uses no outcome feedback. Establishes minimum performance benchmark.
    """

    def initialize(self, arms: List[str], config: Dict[str, Any]) -> None:
        self.arms = arms
        self.step_count = 0
        self.successes: Dict[str, int] = {arm: 0 for arm in arms}
        self.counts: Dict[str, int] = {arm: 0 for arm in arms}

    def select(self, context: TransactionContext) -> str:
        chosen = self.arms[self.step_count % len(self.arms)]
        self.step_count += 1
        return chosen

    def update(self, arm: str, reward: int, context: TransactionContext) -> None:
        self.counts[arm] += 1
        self.successes[arm] += reward

    def get_state(self) -> Dict[str, Any]:
        state = {}
        for arm in self.arms:
            n = self.counts[arm]
            state[arm] = {
                "estimated_sr": self.successes[arm] / n if n > 0 else None,
                "selection_score": None,
                "total_selections": n,
                "total_successes": self.successes[arm],
                "total_failures": n - self.successes[arm],
            }
        return state

    def explain_last_decision(self) -> str:
        idx = (self.step_count - 1) % len(self.arms)
        return f"Round Robin: selected arm #{idx} ('{self.arms[idx]}') at step {self.step_count}."

    def get_hyperparameter_schema(self) -> Dict[str, Any]:
        return {}

    @classmethod
    def metadata(cls) -> Dict[str, str]:
        return {
            "name": "Round Robin",
            "short_name": "RR",
            "description": "Deterministic baseline cycling through all gateways. Uses no outcome feedback. Establishes minimum performance benchmark.",
            "paper": "Deterministic baseline — no research paper",
            "paper_url": "",
            "category": "rule_based",
            "non_stationary": "false",
        }
